app_backend/server_test.py

```diff
@@ -1,9 +1,25 @@
 import server
+import database
 from fastapi.testclient import TestClient
 from fastapi import status
+from dotenv import load_dotenv
+import os
 
 client = TestClient(server.app)
 
+load_dotenv()
+
+class TestDatabase(database.Database):
+    def __init__(self, username, password):
+        super().__init__(username, password)
+
+    def save_user_in_db(self, login_info):
```

Comment:
Could you explain how this works?

Reply:
So the issue with the initial Database code and tests is that you were updating the database live. What that means is that if you run the tests on a clean database everything passes and updates the database appropriately, but when you run the tests again, you get 400 errors because all the 201 codes are getting "email already exists errors". 

During tests, we actually don't want to update the database but rather simulate what would've have happened if we did. That's where `dry_run` comes in. `dry_run` indicates that the code can take a look at the database but it can't touch it. By default this is False so all other code doesn't break, but for tests, we want to overwrite any "save" functions to use this `dry_run` variable. 

Now all tests pass regardless of how many times we run it and we don't mess up any data

Reply:
Okay yeah. I was curious about that when first trying to write the unit tests for database queries. Thanks for finding a good way to do that!